### ISSUE TITLE
New API proposal: square brackets for elementwise function applied to matrix

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#15745]
SymPy currently handles functions of matrices in the mathematical way, for example `exp(M)` is the exponential map of matrix `M`. It would be nice to have support for elementwise application of functions ot matrices.

Recently, `sympy.matrices.expressions.applyfunc.ElementwiseApplyFunction` has been added to SymPy.
https://github.com/sympy/sympy/blob/53b5006a1a56c71fadc954e3ca530879b31db415/sympy/matrices/expressions/applyfunc.py#L5

What about creating a shortcut for this element?

I suggest using `exp[M]` as a shortcut for the elementwise application of `exp` to `M`.

This idea can be easily implemented by adding `__getitem__` to `Function` class.

**Comment by moorepants** [comment]
I don't think that doing non-standard things with hacks to the base Python syntax is good practice. And large widely used libraries like SymPy shouldn't deviate from Python's base syntax. That is, square brackets should get items and parentheses should call functions. Mixing those things up is confusing to users. Why not create a set of function pairs, like `exp()` and `ele_exp()`?

**Comment by Upabjojr** [comment]
> Why not create a set of function pairs, like exp() and ele_exp()?

Shall we create pairs of functions for all functions? This issue is not limited to `exp`.

We already have `M.applyfunc(exp)`, this issue relates to possible extensions to matrix expressions and unevaluated expressions.

**Comment by Upabjojr** [comment]
Will we ever need to overload `__getitem__` for `Function`?

**Comment by moorepants** [comment]
`M.applyfunc(exp)` looks sufficient to me. Maybe make an alias for the method that has a shorter method name.

**Comment by asmeurer** [comment]
I agree with Jason. The other issue is that the syntax would only make sense for a small subset of SymPy types. 

For single functions, applyfunc should work just fine, though I agree we could also make a shorter shortcut for it. The more complicated case would be larger expressions. You'd have to use a lambda/Lambda or elementwise each function independently (and make sure to use HadamardProduct instead of `*`).

**Comment by smichr** [comment]
What about `M(exp)` being a shortcut for `M.applyfunc(exp)`?

**Comment by Upabjojr** [comment]
> What about `M(exp)` being a shortcut for `M.applyfunc(exp)`?

This could create conflicts with matrix functions, if we are going to support them.

**Comment by smichr** [comment]
I don't think the polymorphic input would cause a problem in that the single argument being a function would be interpreted as something to apply to the matrix rather than something the matrix would be applied to.

**Comment by Upabjojr** [comment]
> I don't think the polymorphic input would cause a problem in that the single argument being a function would be interpreted as something to apply to the matrix rather than something the matrix would be applied to.

We can still have functions taking other functions as parameters. Consider Lagrangians in modern physics, they take a field as parameter and may be indexed (stacked in a vector).
